### PR TITLE
Add linear_system_solvers documentation.

### PR DESCRIPTION
**Pull request by Algue-Rythme**
Add documentation for Iterative Refinement with illustrative example of use.

### CODE REVIEW

**Comment by Algue-Rythme** [comment]
Incomplete work: other commits to document content of `jaxopt.linear_solver` will be added soon.

File: docs/linear_system_solvers.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+Linear System Solvers
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:1]:
```suggestion
Linear system solving
```

File: docs/linear_system_solvers.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+.. math::
+
+    Ax = b
+
+with unknown :math:`x` for a linear operator :math:`A` and vector :math:`b`.
+
+
+Iterative Refinement
+--------------------
+
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:15]: Some brief explanation of iterative refinement + link to wikipedia would be great.

File: docs/linear_system_solvers.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  :toctree: _autosummary
+
+    jaxopt.IterativeRefinement
+
+
+Instantiating and running the solver
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The class :class:`IterativeRefinement` aims to improve accuracy of linear solvers 
+from ``jax.scipy.sparse.linalg``.
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:26]: We should probably use our own instead.

File: docs/linear_system_solvers.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    def gmres_ridge(matvec_A, b):
+      def matvec(x):
+        return matvec_A(x) + ridge * x
+    return jax.scipy.sparse.linalg.gmres(matvec, b, tol=tol)[0]
+
+    x_ridge = gmres_ridge(lambda x: jnp.dot(A, x) + ridge * x, b)
+    print(f"GMRES+ridge error: {jnp.linalg.norm(A @ x_ridge - b):.7f}")
+    # GMRES+ridge error: 0.0500162
+
+    solver = IterativeRefinement(solve=gmres_ridge,
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:57]: I'm surprised: why don't we use `solve_gmres` directly?

File: jaxopt/_src/iterative_refinement.py [hunk 5]
```diff
@@ -161,19 +159,30 @@ def update(self,
...
+      init_params: ignored.
+      params_A: params for ``self.matvec_A``.
+      b: vector ``b`` in ``Ax=b``.
+      params_A_bar: optional parameters for ``matvec_A_bar``.
+    Returns:
+      (params, state), ``params = (primal_var, dual_var_eq, dual_var_ineq)``
+    """
+    del init_params
+    # TODO(lbethune): allow warm start when the linear_solve API reaches its final state. 
+    # Currently Jaxopt API uses init keyword but it is not exposed publicly.
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:183]: It is public but not documented (https://github.com/google/jaxopt/blob/main/jaxopt/linear_solve.py).

File: jaxopt/_src/iterative_refinement.py [hunk 6]
```diff
@@ -161,19 +159,30 @@ def update(self,
...
           params_A: Any,
           b: Any,
           params_A_bar: Optional[Any] = None):
-
-    if init_params is None:
-      init_params = self.init_params(params_A, b, params_A_bar)
+    """Runs the iterative refinement.
+
+    Args:
+      init_params: ignored.
```
Comment (mblondel) [inline jaxopt/_src/iterative_refinement.py:174]: It would be better if we don't ignore it.

File: docs/linear_system_solvers.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    tol = 1e-7
+
+    x = jax.scipy.sparse.linalg.gmres(lambda x: jnp.dot(A, x), b, tol=tol)[0]
+    print(f"GMRES only error: {jnp.linalg.norm(A @ x - b):.7f}")
+    # GMRES only error: nan
+
+    def gmres_ridge(matvec_A, b):
+      def matvec(x):
+        return matvec_A(x) + ridge * x
+    return jax.scipy.sparse.linalg.gmres(matvec, b, tol=tol)[0]
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:51]: Indentation issue

**Review by mblondel — CHANGES_REQUESTED** [review]

File: docs/linear_system_solvers.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    x_ridge = gmres_ridge(lambda x: jnp.dot(A, x) + ridge * x, b)
+    print(f"GMRES+ridge error: {jnp.linalg.norm(A @ x_ridge - b):.7f}")
+    # GMRES+ridge error: 0.0500162
+
+    solver = IterativeRefinement(solve=gmres_ridge,
+                                 tol=tol, maxiter=100)
+    x_refined, state = solver.run(init_params=None, params_A=A, b=b)
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:59]: `params_A=A` is a bit confusing, maybe we should rename `params_A` to `A`, although I know `A` is not ideal either in the matvec case...

File: docs/linear_system_solvers.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+This section is concerned with solving problems of the form
+
+.. math::
+
+    Ax = b
+
+with unknown :math:`x` for a linear operator :math:`A` and vector :math:`b`.
+
+
+Iterative Refinement
```
Comment (mblondel) [inline docs/linear_system_solvers.rst:13]:
```suggestion
Iterative refinement
```

File: docs/linear_system_solvers.rst (hunk 4 above)
Comment (Algue-Rythme) [reply docs/linear_system_solvers.rst:57]: That was written before you confirmed that you want to see `jaxopt.linear_solve` in documentation; I thought I couldn't use undocumentated code for the example. I will switch to `jaxopt.linear_solve` now you confirmed it was ok to do so.

**Pull request merged**